File: app/src/main/java/ethiopia/covid/android/ui/adapter/CheckBoxQuestionRecyclerAdapter.java
package ethiopia.covid.android.ui.adapter;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CompoundButton;

import androidx.annotation.NonNull;
import androidx.appcompat.widget.AppCompatTextView;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.checkbox.MaterialCheckBox;

import java.util.List;

import ethiopia.covid.android.R;
import ethiopia.covid.android.data.QuestionItem;

/**
 * Created by BrookMG on 4/12/2020 in ethiopia.covid.android.ui.adapter
 * inside the project CoVidEt .
 */
public class CheckBoxQuestionRecyclerAdapter extends RecyclerView.Adapter<CheckBoxQuestionRecyclerAdapter.ViewHolder> {

    private List<QuestionItem> questionItems;
    private OnQuestionItemSelected onQuestionCheckBoxClicked;

    public CheckBoxQuestionRecyclerAdapter(List<QuestionItem> questionItems, OnQuestionItemSelected onQuestionCheckBoxClicked) {
        this.questionItems = questionItems;
        this.onQuestionCheckBoxClicked = onQuestionCheckBoxClicked;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new ViewHolder(
                LayoutInflater.from(parent.getContext())
                        .inflate(R.layout.question_page_two_recycler_element , parent, false)
        );
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        holder.mainCheckBox.setText(questionItems.get(position).getQuestionText());
        holder.mainCheckBox.setOnCheckedChangeListener((buttonView, isChecked) -> {
            if (onQuestionCheckBoxClicked != null) {
                onQuestionCheckBoxClicked.onItemSelected(
                        questionItems.get(holder.getAdapterPosition()),
                        holder.getAdapterPosition()
                );
            }
        });
    }

    @Override
    public int getItemCount() {
        return questionItems.size();
    }

    static class ViewHolder extends RecyclerView.ViewHolder {

        MaterialCheckBox mainCheckBox;

        public ViewHolder(@NonNull View itemView) {
            super(itemView);
            mainCheckBox = itemView.findViewById(R.id.main_check_box);
        }
    }

}

File: app/src/main/java/ethiopia/covid/android/ui/adapter/CheckBoxQuestionRecyclerAdapter.kt
package ethiopia.covid.android.ui.adapter

import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.CompoundButton
import androidx.recyclerview.widget.RecyclerView
import ethiopia.covid.android.data.QuestionItem
import ethiopia.covid.android.databinding.QuestionPageTwoRecyclerElementBinding

/**
 * Created by BrookMG on 4/12/2020 in ethiopia.covid.android.ui.adapter
 * inside the project CoVidEt .
 */
class CheckBoxQuestionRecyclerAdapter(
        private val questionItems: List<QuestionItem>,
        private val singleCheckBoxOnly: Boolean,
        private val onQuestionCheckBoxClicked: OnQuestionItemSelected?
) : RecyclerView.Adapter<CheckBoxQuestionRecyclerAdapter.ViewHolder>() {

    constructor(questionItems: List<QuestionItem>, onQuestionCheckBoxClicked: OnQuestionItemSelected?)
            : this(questionItems, false, onQuestionCheckBoxClicked)

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder =
            ViewHolder(
                    QuestionPageTwoRecyclerElementBinding.inflate(
                            LayoutInflater.from(parent.context), parent, false
                    ), this
            )

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(questionItems, questionItems[position], singleCheckBoxOnly, onQuestionCheckBoxClicked)
    }

    override fun getItemCount(): Int = questionItems.size

    class ViewHolder(
            private val questionPageTwoRecyclerElementBinding: QuestionPageTwoRecyclerElementBinding,
            private val recyclerAdapter: CheckBoxQuestionRecyclerAdapter
    ) : RecyclerView.ViewHolder(questionPageTwoRecyclerElementBinding.root) {

        private fun singleSelectItem(questionItems: List<QuestionItem>, position: Int, selection: Boolean) {
            for (i in questionItems.indices) {
                if (i == position) questionItems[position].isSelectedQuestion = selection
                else questionItems[i].isSelectedQuestion = false
                questionPageTwoRecyclerElementBinding.root.post {
                    recyclerAdapter.notifyItemChanged(i)
                }
            }
        }

        fun bind(questionItems: List<QuestionItem>, questionItem: QuestionItem, singleCheckBoxOnly: Boolean,
                 onQuestionCheckBoxClicked: OnQuestionItemSelected?) {
            questionPageTwoRecyclerElementBinding.let {
                it.mainCheckBox.text = questionItem.questionText
                it.mainCheckBox.setOnCheckedChangeListener(null)
                it.mainCheckBox.isChecked = questionItem.isSelectedQuestion

                it.mainCheckBox.setOnCheckedChangeListener { _: CompoundButton?, isChecked: Boolean ->
                    if (singleCheckBoxOnly) {
                        singleSelectItem(questionItems, adapterPosition, isChecked)
                    } else {
                        questionItems[adapterPosition].isSelectedQuestion = isChecked
                    }

                    onQuestionCheckBoxClicked?.onItemSelected(
                            questionItems[adapterPosition],
                            adapterPosition
                    )
                }
            }
        }

    }

}